config/configtls/configtls.go

```diff
@@ -104,14 +109,26 @@ func (c TLSSetting) loadTLSConfig() (*tls.Config, error) {
 		return nil, fmt.Errorf("for auth via TLS, either both certificate and key must be supplied, or neither")
 	}
 
-	var certificates []tls.Certificate
+	var getCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
 	if c.CertFile != "" && c.KeyFile != "" {
-		var tlsCert tls.Certificate
-		tlsCert, err = tls.LoadX509KeyPair(filepath.Clean(c.CertFile), filepath.Clean(c.KeyFile))
+		lastReload := time.Now()
+		var cert tls.Certificate
+		cert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
 		if err != nil {
 			return nil, fmt.Errorf("failed to load TLS cert and key: %w", err)
 		}
-		certificates = append(certificates, tlsCert)
+		getCertificate = func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			now := time.Now()
+			if c.ReloadInterval != 0 && lastReload.Add(c.ReloadInterval).Before(now) {
+				var newCert tls.Certificate
+				newCert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
+				if err == nil {
```

Comment:
If there was an error loading the new cert it should be logged.

Reply:
I would agree, but I can use some guidance on how you'd prefer to pass in a logger here: we need access to the logger when config is evaluated so it can be put into the closure.

An alternative is to just return the error out of `GetCertificates`: this will cause the TCP connection to fail, but will log the error we return here.
The reason I didn't is because the most likely reason for failure is that the new certs are in the process of being written, which is transient and will typically succeed afterwards.

Reply:
If a returned error gets logged that may be preferable.  Failed connections can and should be retried and, if the failure is indeed transient, then it will soon succeed.  But if the failure isn't transient things will eventually start failing with expired or otherwise invalid certs, all the while the user is expecting that the new cert is being used when it isn't.